contracts/contracts/FeeSub.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/token/ERC20/IERC20.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/token/ERC20/utils/SafeERC20.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.8.0/contracts/access/Ownable.sol";
+
+import "./RequestManager.sol";
+
+contract FeeSub is Ownable {
+    using SafeERC20 for IERC20;
+
+    RequestManager public requestManager;
+    mapping(address token => uint256) public minimumAmounts;
+    mapping(bytes32 requestId => address sender) public senders;
+
+    constructor(address _requestManager) {
+        requestManager = RequestManager(_requestManager);
+    }
+
+    function createRequest(
+        uint256 targetChainId,
+        address sourceTokenAddress,
+        address targetTokenAddress,
+        address targetAddress,
+        uint256 amount,
+        uint256 validityPeriod
+    ) external returns (bytes32) {
+        require(
+            minimumAmounts[sourceTokenAddress] > 0,
+            "Token not to be subsidized"
+        );
+        require(
+            amount >= minimumAmounts[sourceTokenAddress],
+            "Transfer amount too small to be subsidized"
+        );
+
+        require(
+            IERC20(sourceTokenAddress).allowance(msg.sender, address(this)) >=
+                amount,
+            "Insufficient allowance"
+        );
+
+        IERC20(sourceTokenAddress).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amount
+        );
+
+        bytes32 requestId = requestManager.createRequest(
+            targetChainId,
+            sourceTokenAddress,
+            targetTokenAddress,
+            targetAddress,
+            amount,
+            validityPeriod
+        );
+        senders[requestId] = msg.sender;
+
+        return requestId;
+    }
+
+    function transferableAmount(
+        uint256 targetChainId,
+        address tokenAddress,
+        uint256 amount
+    ) public view returns (uint256) {
+        if (amount >= minimumAmounts[tokenAddress]) {
+            return amount;
+        }
+
+        return
+            requestManager.transferableAmount(
+                targetChainId,
+                tokenAddress,
+                amount
+            );
+    }
+
+    function withdrawExpiredRequest(bytes32 requestId) external {
+        (
+            address sender,
+            address sourceTokenAddress,
+            ,
+            uint256 amount,
+            ,
+            ,
+            ,
+            ,
+            ,
+            ,
+
+        ) = requestManager.requests(requestId);
+        require(
+            sender == address(this),
+            "Request was not created by this contract"
+        );
+        require(senders[requestId] != address(0), "Already Withdrawn");
+
+        requestManager.withdrawExpiredRequest(requestId);
```

Comment:
Here we need to add another condition. It can be the case that the request was already withdrawn by calling the RequestManager directly. 
If that's the case this call would revert and the user's money would be stuck inside this contract forever. 

I suggest that we do this check by checking the withdrawnClaimId of the request. 
We should only call `requestManager.withdrawExpiredRequest(requestId);` if withdrawnClaimId is 0.

Reply:
Speaking of which, we also need an additional require that the tokens have not been withdrawn by the agent. 
IOW, we should only ever proceed if we either can call `withdrawExpiredRequest` or we are sure that the money was already withdrawn, but to this contract


Suggestion: 

```sol
if(withdrawClaimId == 0) {
    // This will fail if the funds do not belong to the user (yet)
    requestManager.withdrawExpiredRequest(requestId);
}
else {
    // Make sure that funds were withdrawn by calling withdrawExpiredRequest()
    require(withdrawnClaimId == type(uint96).max);

}


```